### PR TITLE
Add per-op MPS gradient tests and update skips

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84242

Follow up:
- ~Remove non-float dtypes from allow-list for gradients~
- ~Map dtypes to short-hand so there aren't so many lines, i.e. float16 should be f16.~
- ~There were a lot of linting issues that flake8 wouldn't format for me, so I reformatted with black. This makes the diff a little trickier to parse.~

Observations:
- there are entries in the allow-list that weren't there before 
- some forward that we previously passing now fail with requires_grad=True
- Because the allow list does not know about variants, a special skip was added for that in the block list

### CODE REVIEW

File: test/test_mps.py [hunk 1]
```diff
@@ -6405,543 +6406,3244 @@ class TestConsistency(TestCase):
...
-                   'torch.int16',
-                   'torch.int32',
-                   'torch.int64'],
-        'zero_': ['torch.float16',
-                  'torch.float32',
-                  'torch.int16',
-                  'torch.int32',
-                  'torch.int64',
-                  'torch.uint8']}
+        "__getitem__": [
```
Comment (albanD) [inline test/test_mps.py:6409]: I feel like these either need to become blocklist (as there are more enabled than disabled) or we should split it in a different file to keep this one at a reasonable size.

File: test/test_mps.py [hunk 2]
```diff
@@ -6962,18 +9664,42 @@ def test_output_match(self, device, dtype, op):
...
+                or str(dtype) not in self.ALLOWLIST_OP_GRAD[op.name]
+            ):
+                run_grad_test = False
+
+        def get_samples():
+            return op.sample_inputs(
+                device, dtype, requires_grad=(dtype.is_floating_point or dtype.is_complex)
+            )
+        cpu_samples = get_samples()
+        nb_samples = len(list(get_samples()))
```
Comment (albanD) [inline test/test_mps.py:9688]: Why?

File: test/test_mps.py [hunk 3]
```diff
@@ -6985,18 +9711,107 @@ def test_output_match(self, device, dtype, op):
                 cpu_out = op(*cpu_args, **cpu_kwargs)
                 mps_out = op(*mps_args, **mps_kwargs)
                 self.assertEqual(cpu_out, mps_out)
-        except Exception as e:
-            if not generate_new_truth:
-                raise e
-        else:
```
Comment (albanD) [inline test/test_mps.py:6991]: Note that this is not in the except, this is in the else that we append to the allow list.

File: test/test_mps.py [hunk 4]
```diff
@@ -6985,18 +9711,107 @@ def test_output_match(self, device, dtype, op):
...
-                # We could write it only once. But I don't know how to detect that the current test is the last one
-                # So each test append to the dict and write it.
+            except Exception as e:
+                if not generate_new_truth:
+                    raise e
+                forward_failed = True
+                all_forward_pass = False
+
+            # Skip the grad test if it is not part of the allow list
+            if not generate_new_truth and not run_grad_test:
```
Comment (albanD) [inline test/test_mps.py:9722]: This would better within the "Backward check" section I think.

File: test/test_mps.py [hunk 5]
```diff
@@ -6985,18 +9711,107 @@ def test_output_match(self, device, dtype, op):
...
+                if i == 0:
```
Comment (albanD) [inline test/test_mps.py:9723]: debug print

File: test/test_mps.py [hunk 6]
```diff
@@ -6985,18 +9711,107 @@ def test_output_match(self, device, dtype, op):
...
+                    t
+                    for t in pytree.tree_flatten((cpu_args, cpu_kwargs))[0]
+                    if req_grad(t)
+                )
+                diff_mps_arg = tuple(
+                    t
+                    for t in pytree.tree_flatten((mps_args, mps_kwargs))[0]
+                    if req_grad(t)
+                )
+                self.assertEqual(len(diff_cpu_out), len(diff_mps_out))
```
Comment (albanD) [inline test/test_mps.py:9763]: Also check input match as a sanity check?

File: test/test_mps.py [hunk 7]
```diff
@@ -6985,18 +9711,107 @@ def test_output_match(self, device, dtype, op):
...
+            # Skip the grad test if it is not part of the allow list
+            if not generate_new_truth and not run_grad_test:
+                if i == 0:
+                    print(
+                        f"Skipping gradient check because {op.name} is not on the allow list"
+                    )
+                continue
+
+            if not (dtype.is_floating_point or dtype.is_complex):
+                continue
```
Comment (albanD) [inline test/test_mps.py:9730]: This means that you never generate forward new truth for non-floating/complex dtypes?

File: test/test_mps.py [hunk 8]
```diff
@@ -6985,18 +9711,107 @@ def test_output_match(self, device, dtype, op):
...
+                if not generate_new_truth:
+                    raise e
+                all_backward_pass = False
+
+        if all_forward_pass and generate_new_truth:
+            if str(dtype) not in self.NEW_ALLOW_LIST[op.name]:
+                self.NEW_ALLOW_LIST[op.name].append(str(dtype))
+            # We could write it only once. But I don't know how to detect that the current test is the last one
+            # So each test append to the dict and write it.
+            if i >= nb_samples - 1:
```
Comment (albanD) [inline test/test_mps.py:9801]: == ?

**Review by albanD — COMMENTED** [review]
Sounds pretty good!

File: test/test_mps.py (hunk 2 above)
Comment (soulitzer) [reply test/test_mps.py:7017]: Whoops, forgot to clean this up

File: test/test_mps.py (hunk 8 above)
Comment (soulitzer) [reply test/test_mps.py:7101]: Yep, but no longer needed

File: test/test_mps.py (hunk 3 above)
Comment (soulitzer) [reply test/test_mps.py:6991]: Yup, I later realized that too lol

File: test/test_mps.py (hunk 1 above)
Comment (soulitzer) [reply test/test_mps.py]: I've updated to use abbreviations for the dtypes (aka f16 instead of torch.float16) to reduce the size, perhaps it is more reasonable now?

**Comment by kulinseth** [comment]
Overall the commit looks good. Thanks for adding these new tests @soulitzer . I added the ciflow/mps labels to run the test_mps to make sure its green.

**Review by kulinseth — APPROVED** [review]

**Review by malfet — APPROVED** [review]
